config/src/config/config_optimizer.rs

```diff
@@ -0,0 +1,219 @@
+// Copyright © Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    config::{
+        node_config_loader::NodeType, utils::get_config_name, Error, InspectionServiceConfig,
+        LoggerConfig, NodeConfig, StateSyncConfig,
+    },
+    network_id::NetworkId,
+};
+use aptos_types::chain_id::ChainId;
+use serde_yaml::Value;
+
+// Useful optimizer constants
+const OPTIMIZER_STRING: &str = "Optimizer";
+const VALIDATOR_NETWORK_OPTIMIZER_NAME: &str = "ValidatorNetworkConfigOptimizer";
+
+/// A trait for optimizing node configs (and their sub-configs) by tweaking
+/// config values based on node types, chain IDs and compiler features.
+///
+/// Note: The config optimizer respects the following order precedence when
+/// determining whether or not to optimize a value:
+/// 1. If a config value has been set in the local config file, that value
+///    should be used (and the optimizer should not override it).
+/// 2. If a config value has not been set in the local config file, the
+///    optimizer may set the value (but, it is not required to do so).
+/// 3. Finally, if the config optimizer chooses not to set a value, the default
+///    value is used (as defined in the default implementation).
+pub trait ConfigOptimizer {
```

Comment:
is there a clear split between 'config optimizer' (sounds like 'makes config better') and 'config sanitizer' (sounds like 'makes config correct') ? should they just be one thing?

Reply:
Yeah, they are entirely separate config passes:
- Optimizer defines fine-grained defaults
- Sanitizer ensures properties/invariants hold and will throw errors if they don't.

Technically, the sanitizer should never modify the config (only the optimizer).